src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs

```diff
@@ -810,20 +828,45 @@ public Task SendMailAsync(MailMessage message)
                 throw;
             }
 
+            // Only register on the CT if HandleCompletion hasn't started to ensure the CTR is disposed
+            bool lockTaken = false;
+            try
+            {
+                Monitor.TryEnter(tcs, ref lockTaken);
+                if (lockTaken && !tcs.Task.IsCompleted)
+                {
+                    ctr = cancellationToken.Register(s =>
+                    {
+                        ((SmtpClient)s).SendAsyncCancel();
```

Comment:
I'm not familiar with the SmtpClient class, but what does `SendAsyncCancel()` do? It would seem like it's cancelling every ongoing async operation. Does that mean that other concurrent send operations that do not specify a cancellation token would get cancelled too?

Reply:
SmtpClient is not threadsafe, only one request can be active at a time.

SendAsyncCancel tries to abort a currently ongoing send operation, but due to poor legacy api design that isn't actually a guarantee (if called right after SendAsync). That is why the CTR is registered only after calling SendAsync and only if the request hasn't finished.